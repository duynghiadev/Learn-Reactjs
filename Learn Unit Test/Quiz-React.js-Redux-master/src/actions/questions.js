// Node modules import
import axios from 'axios';

// Actions types import
import {
  FETCH_QUESTION,
  DELETE_CURRENT_QUESTION,
  INCREMENT_TOTAL_COUNT,
  INCREMENT_CORRECT_COUNT,
  RELOCATE_FROM_PROPOSITION_TO_BOARD,
  RELOCATE_FROM_BOARD_TO_PROPOSITION,
  CHECK_ANSWER,
  CHANGE_SCORE,
  SHOW_ERROR,
  HIDE_ERROR
 } from '../constants/questions.js';

// Functions import
import { stringToArray, shuffleArray } from '../helpers/functions.js'

// Receives a random question
export function fetchQuestion() {
  return function(dispatch) {
    return axios.get('http://jservice.io/api/random')
      .then(function(response) {
        // TODO: maybe delete this line
        const arrayOfCharacters = stringToArray(response.data[0].answer);
        const shuffledArray = shuffleArray(arrayOfCharacters);
        const data = {
          id: response.data[0].id,
          category: response.data[0].category,
          question: response.data[0].question,
          answer: response.data[0].answer,
          value: response.data[0].value,
          transformedAnswer: shuffledArray
        }
        dispatch({
          type: FETCH_QUESTION,
          payload: data
        })
      }, function(error) {
        dispatch({
          type: SHOW_ERROR
        })
      })
  }
}

// Hides an error and allowes to continue use the application
export function hideError() {
  return function(dispatch) {
    dispatch({
      type: HIDE_ERROR
    })
  }
}

// Delete current question (to see the loader)
export function removeQuestion() {
  return function(dispatch) {
    dispatch({
      type: DELETE_CURRENT_QUESTION
    })
  }
}

// Increments the total questions count
export function incrementTotalCount() {
  return function(dispatch) {
    dispatch({
      type: INCREMENT_TOTAL_COUNT
    })
  }
}

// Increments the correct questions count
export function incrementCorrectCount() {
  return function(dispatch) {
    dispatch({
      type: INCREMENT_CORRECT_COUNT
    })
  }
}

// Increments/decrements the score by the given amount
export function changeScore(amount) {
  return function(dispatch) {
    dispatch({
      type: CHANGE_SCORE,
      payload: amount
    })
  }
}

// Relocates character from  "charsInProposition" to "charsOnBoard"
export function relocateToBoard(char) {
  return function(dispatch) {
    dispatch({
      type: RELOCATE_FROM_PROPOSITION_TO_BOARD,
      payload: char
    })
  }
}

// Relocates character from  "charsOnBoard" to "charsInProposition"
export function relocateToProposition(char) {
  return function(dispatch) {
    dispatch({
      type: RELOCATE_FROM_BOARD_TO_PROPOSITION,
      payload: char
    })
  }
}

// Checks if the answer provided by user is correct
export function checkAnswer(char) {
  return function(dispatch) {
    dispatch({
      type: CHECK_ANSWER,
      payload: char
    })
  }
}
