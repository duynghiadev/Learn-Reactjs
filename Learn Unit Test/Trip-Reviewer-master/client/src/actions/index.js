import axios from 'axios';

import {
  GET_TRIPS,
  GET_USERS,
  GET_TRIP_W_REVIEWER,
  CLEAR_TRIP_W_REVIEWER,
  GET_TRIP_W_REVIEWER_ERROR,
  USER_REGISTER,
  USER_LOGIN,
  RESET_PASSWORD,
  USER_AUTH,
  ADD_TRIP,
  UPDATE_TRIP,
  UPDATE_USER,
  DELETE_TRIP,
  GET_TRIP,
  GET_TRIP_ERROR,
  CLEAR_NEW_TRIP,
  CLEAR_TRIP,
  GET_USER_REVIEWS,
} from '../constants/action-names';

export function getTrips(limit = 10, start = 0, order = 'asc', list = []) {
  const request = axios.get(`/api/getManyTrips?limit=${limit}&skip=${start}&order=${order}`);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: GET_TRIPS,
        payload: {
          list: [...list, ...response.data],
          newTripsCount: response.data.length,
        },
      });
    });
}

export function getTripWithReviewer(id) {
  const request = axios.get(`/api/getTrip?id=${id}`);

  return dispatch =>
    request
      .then(({ data: tripData }) => {
        const trip = tripData;

        return axios.get(`/api/getReviewer?id=${trip.ownerId}`).then(({ data: reviewerData }) => {
          const response = {
            trip,
            reviewer: reviewerData,
          };

          dispatch({
            type: GET_TRIP_W_REVIEWER,
            payload: response,
          });
        });
      })
      .catch(error =>
        dispatch({
          type: GET_TRIP_W_REVIEWER_ERROR,
          payload: error,
        })
      );
}

export function getUserReviews(userId) {
  const request = axios.get(`/api/getUserReviews?user=${userId}`);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: GET_USER_REVIEWS,
        payload: response.data,
      });
    });
}

export function updateUser(data) {
  const request = axios.post('/api/userUpdate', data);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: UPDATE_USER,
        payload: response.data,
      });
    });
}

export function getTrip(id) {
  const request = axios.get(`/api/getTrip?id=${id}`);

  return dispatch =>
    request
      .then(response => {
        dispatch({
          type: GET_TRIP,
          payload: response.data,
        });
      })
      .catch(error => {
        dispatch({
          type: GET_TRIP_ERROR,
          payload: error,
        });
      });
}

export function updateTrip(data) {
  const request = axios.post('/api/tripUpdate', data);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: UPDATE_TRIP,
        payload: response.data,
      });
    });
}

export function deleteTrip(id) {
  const request = axios.delete(`/api/tripDelete?id=${id}`);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: DELETE_TRIP,
        payload: response.data,
      });
    });
}

export function clearTrip() {
  return {
    type: CLEAR_TRIP,
    payload: {
      trip: null,
      updatedTrip: false,
      postDeleted: false,
    },
  };
}

export function clearTripWithReviewer() {
  return {
    type: CLEAR_TRIP_W_REVIEWER,
    payload: {
      trip: null,
      reviewer: null,
    },
  };
}

export function loginUser({ email, password }) {
  const request = axios.post('/api/login', { email, password });

  return dispatch =>
    request.then(response => {
      dispatch({
        type: USER_LOGIN,
        payload: response.data,
      });
    });
}

export function resetPassword({ email }) {
  const request = axios.post('/api/resetPassword', { email });

  return dispatch =>
    request.then(response => {
      dispatch({
        type: RESET_PASSWORD,
        payload: response.data,
      });
    });
}

export function auth() {
  const request = axios.get('/api/auth');

  return dispatch =>
    request.then(response => {
      dispatch({
        type: USER_AUTH,
        payload: response.data,
      });
    });
}

export function addTrip(trip) {
  const request = axios.post('/api/trip', trip);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: ADD_TRIP,
        payload: response.data,
      });
    });
}

export function clearNewTrip() {
  return {
    type: CLEAR_NEW_TRIP,
    payload: null,
  };
}

export function getUsers() {
  const request = axios.get('/api/users');

  return dispatch =>
    request.then(response => {
      dispatch({
        type: GET_USERS,
        payload: response.data,
      });
    });
}

export function userRegister(user) {
  const request = axios.post('/api/register', user);

  return dispatch =>
    request.then(({ data }) => {
      dispatch({
        type: USER_REGISTER,
        payload: data,
      });
    });
}
