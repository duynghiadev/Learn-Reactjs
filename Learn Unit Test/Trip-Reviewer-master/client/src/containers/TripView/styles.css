.trip-view-container {
  box-sizing: border-box;
  padding: 20px 20px 50px 20px;
}
.trip-view-container .trip-view-header {
  border-bottom: 1px solid #e1e1e1;
  margin-bottom: 20px;
}
.trip-view-container .trip-view-header h2 {
  font-weight: 300;
  font-size: 38px;
  margin: 0px;
}
.trip-view-container .trip-view-header h5 {
  margin: 5px 0px;
  font-size: 31px;
  color: #02bcd4;
}
.trip-view-container .trip-view-header .reviewer {
  font-size: 14px;
  margin-bottom: 10px;
}
.trip-view-container .trip-view-header .reviewer span {
  font-weight: 500;
}
.trip-view-container .trip-view-description {
  font-size: 20px;
  font-weight: 300;
  line-height: 27px;
}
.trip-view-container .trip-view-box {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 40px;
}
.trip-view-container .trip-view-box .left {
  min-width: 50%;
  background: #e91e63;
  color: #fff;
  box-sizing: border-box;
  padding: 15px;
  font-weight: 300;
  font-size: 20px;
}
.trip-view-container .trip-view-box .left div {
  padding: 7px;
  box-sizing: border-box;
}
.trip-view-container .trip-view-box .left div span {
  font-weight: 700;
}
.trip-view-container .trip-view-box .right {
  flex-grow: 1;
  height: 106px;
  background: #3c3c3c;
  color: #fff;
  font-weight: 700;
  padding: 21px 15px;
  box-sizing: border-box;
  text-align: center;
}
.trip-view-container .trip-view-box .right div {
  font-size: 40px;
  font-weight: 300;
}

.trip-view-container .editor-toolbar {
  display: none;
}
