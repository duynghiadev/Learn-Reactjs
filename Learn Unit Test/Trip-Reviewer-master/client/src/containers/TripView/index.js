import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Editor } from 'react-draft-wysiwyg';
import { EditorState, convertFromRaw } from 'draft-js';

import { getTripWithReviewer, clearTripWithReviewer } from '../../actions';

import './styles.css';

class TripView extends Component {
  componentWillMount() {
    this.props.dispatch(getTripWithReviewer(this.props.match.params.id));
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.trips && nextProps.trips.error) {
      this.props.history.push('/Not-Found');
    }
  }

  componentWillUnmount() {
    this.props.dispatch(clearTripWithReviewer());
  }

  getEditorState = state => {
    let editorState = EditorState.createEmpty();

    try {
      const parsedDescription = JSON.parse(state);
      const contentState = convertFromRaw(parsedDescription);
      editorState = EditorState.createWithContent(contentState);
    } catch (e) {
      console.log('Parsing description error: ', e); // eslint-disable-line no-console
      console.log('Object: ', state); // eslint-disable-line no-console
    }

    return editorState;
  };

  render() {
    const { trips } = this.props;

    if (trips && trips.current) {
      const { title, country, description, duration, rating, expences } = trips.current;
      const { name, lastname } = trips.reviewer;

      return (
        <div className="trip-view-container limited-width-shadow">
          <div className="trip-view-header">
            <h2>{title}</h2>
            <h5>{country}</h5>
            <div className="reviewer">
              <span>Author:</span>
              {` ${name} ${lastname}`}
            </div>
          </div>
          <div className="trip-view-description">
            <Editor
              editorState={this.getEditorState(description)}
              wrapperClassName="editor-wrapper"
              toolbarClassName="editor-toolbar"
              editorClassName="editor-self"
              toolbarHidden
              readOnly
            />
          </div>
          <div className="trip-view-box">
            <div className="left">
              <div>
                <span>Duration:</span>
                {duration === 1 ? ' 1 day' : ` ${duration} days`}
              </div>
              <div>
                <span>Expenses:</span>
                {` $${expences} `}
              </div>
            </div>
            <div className="right">
              <span>Rating</span>
              <div>{rating}/5</div>
            </div>
          </div>
        </div>
      );
    }
    return null;
  }
}

TripView.propTypes = {
  trips: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    trips: state.trips,
  };
}

export default connect(mapStateToProps)(TripView);
