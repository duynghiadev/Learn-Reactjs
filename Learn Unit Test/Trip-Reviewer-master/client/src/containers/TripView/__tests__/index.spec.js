import React from 'react';
import configureStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import TripView from '..';
import { clearTripWithReviewer } from '../../../actions';

jest.mock('react-draft-wysiwyg', () => ({ Editor: 'Editor' }));
jest.mock('draft-js', () => ({
  EditorState: {
    createEmpty: () => ({}),
    createWithContent: () => ({}),
  },
  convertFromRaw: () => ({}),
}));
jest.mock('./../../../actions', () => ({
  getTripWithReviewer: jest.fn(() => ({
    type: 'getTripWithReviewer',
  })),
  clearTripWithReviewer: jest.fn(() => ({
    type: 'clearTripWithReviewer',
  })),
}));

const mockStore = configureStore();

const mockComponent = (initialState = {}, props) => {
  const store = mockStore(initialState);

  return <TripView {...props} store={store} />;
};

describe('<TripView />', () => {
  it('should render component with empty trips', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };

    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with current trip of 1 day and correct description', () => {
    const initialState = {
      trips: {
        current: {
          title: 'title',
          country: 'country',
          description: '{"test":"test"}',
          duration: 1,
          expences: 'expences',
          rating: 'rating',
        },
        reviewer: {
          name: 'name',
          lastname: 'lastname',
        },
      },
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with current trip of 10 days and wrong', () => {
    const initialState = {
      trips: {
        current: {
          title: 'title',
          country: 'country',
          description: 'description',
          duration: 10,
          expences: 'expences',
          rating: 'rating',
        },
        reviewer: {
          name: 'name',
          lastname: 'lastname',
        },
      },
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should dispatch clearTripWithReviewer when componentWillUnmount is called', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillUnmount();

    expect(clearTripWithReviewer).toHaveBeenCalled();
  });

  it('should redirect when componentWillReceiveProps is called with error', () => {
    const push = jest.fn();
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
      history: { push },
    };
    const newProps = {
      trips: { error: true },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(newProps);

    expect(push).toHaveBeenCalledWith('/Not-Found');
  });

  it('should not redirect when componentWillReceiveProps is called without error', () => {
    const push = jest.fn();
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
      history: { push },
    };
    const newProps = {
      trips: { error: false },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(newProps);

    expect(push).not.toHaveBeenCalledWith('/Not-Found');
  });
});
