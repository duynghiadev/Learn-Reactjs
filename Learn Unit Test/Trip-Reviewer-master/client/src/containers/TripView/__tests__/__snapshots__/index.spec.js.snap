// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripView /> should render component with current trip of 1 day and correct description 1`] = `
<div
  className="trip-view-container limited-width-shadow"
>
  <div
    className="trip-view-header"
  >
    <h2>
      title
    </h2>
    <h5>
      country
    </h5>
    <div
      className="reviewer"
    >
      <span>
        Author:
      </span>
       name lastname
    </div>
  </div>
  <div
    className="trip-view-description"
  >
    <Editor
      editorClassName="editor-self"
      editorState={Object {}}
      readOnly={true}
      toolbarClassName="editor-toolbar"
      toolbarHidden={true}
      wrapperClassName="editor-wrapper"
    />
  </div>
  <div
    className="trip-view-box"
  >
    <div
      className="left"
    >
      <div>
        <span>
          Duration:
        </span>
         1 day
      </div>
      <div>
        <span>
          Expenses:
        </span>
         $expences 
      </div>
    </div>
    <div
      className="right"
    >
      <span>
        Rating
      </span>
      <div>
        rating
        /5
      </div>
    </div>
  </div>
</div>
`;

exports[`<TripView /> should render component with current trip of 10 days and wrong 1`] = `
<div
  className="trip-view-container limited-width-shadow"
>
  <div
    className="trip-view-header"
  >
    <h2>
      title
    </h2>
    <h5>
      country
    </h5>
    <div
      className="reviewer"
    >
      <span>
        Author:
      </span>
       name lastname
    </div>
  </div>
  <div
    className="trip-view-description"
  >
    <Editor
      editorClassName="editor-self"
      editorState={Object {}}
      readOnly={true}
      toolbarClassName="editor-toolbar"
      toolbarHidden={true}
      wrapperClassName="editor-wrapper"
    />
  </div>
  <div
    className="trip-view-box"
  >
    <div
      className="left"
    >
      <div>
        <span>
          Duration:
        </span>
         10 days
      </div>
      <div>
        <span>
          Expenses:
        </span>
         $expences 
      </div>
    </div>
    <div
      className="right"
    >
      <span>
        Rating
      </span>
      <div>
        rating
        /5
      </div>
    </div>
  </div>
</div>
`;

exports[`<TripView /> should render component with empty trips 1`] = `null`;
